### PR TITLE
Address use-def violations arising due to index computation when rewriting tensor.load/tensor.store ops.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
When creating `flow.dispatch.tensor.load`/`flow.dispatch.tensor.store` based on already existing operations (during
`-iree-codegen-tile-and-distribute-to-workgroups` and `-iree-convert-to-destination-passing-style`), using the `Value`s for the offsets, sizes and strides directly can lead to SSA use-def violations. So clone the computation of these values. These are expected to be a few operations that eventually get CSEed.  Add helper methods to do this in `Codegen/Transforms.h` and
`Codegen/Transforms.cpp` for general use.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Transforms/Transforms.h [hunk 1]
```diff
@@ -23,16 +24,33 @@
...
+                                 SmallVector<OpFoldResult> &sizes,
+                                 SmallVector<OpFoldResult> &strides,
+                                 SmallVector<Value> &dynamicDims);
+void cloneOffsetsSizesAndStrides(OpBuilder &builder,
+                                 IREE::Flow::DispatchTensorLoadOp loadOp,
+                                 SmallVector<OpFoldResult> &offsets,
+                                 SmallVector<OpFoldResult> &sizes,
+                                 SmallVector<OpFoldResult> &strides,
+                                 SmallVector<Value> &dynamicDims);
+
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Transforms/Transforms.h:43]: should we move these to https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Codegen/Utils/Utils.h ?

Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Transforms/Transforms.h:43]: I'd rather keep these in `Transforms` cause it is actually changing code... I think `Utils` is just for analysis/querrying methods.

Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/Transforms/Transforms.h:43]: I was confused about what to put into `Utils`. Maybe we can rename it to `AnalysisUtils.h`, I can do the cleanup.

File: compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+                                 IREE::Flow::DispatchTensorLoadOp loadOp,
+                                 SmallVector<OpFoldResult> &offsets,
+                                 SmallVector<OpFoldResult> &sizes,
+                                 SmallVector<OpFoldResult> &strides,
+                                 SmallVector<Value> &dynamicDims) {
+  return cloneOffsetsSizesAndStridesImpl(
+      builder, loadOp, ValueRange{loadOp.getSource()}, loadOp.getMixedOffsets(),
+      loadOp.getMixedSizes(), loadOp.getMixedStrides(), loadOp.getSourceDims(),
+      offsets, sizes, strides, dynamicDims);
+}
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp:92]: These function takes tooo many arguments. I think we can use upstream struct `Range` in `mlir/Dialect/Utils/StaticValueUtils.h`.

```cpp
/// Represents a range (offset, size, and stride) where each element of the
/// triple may be dynamic or static.
struct Range {
  OpFoldResult offset;
  OpFoldResult size;
  OpFoldResult stride;
};
```

and we rewrite it to return a `SmallVector<Range>`. I think it is a trivial case for [RVO](https://en.wikipedia.org/wiki/Copy_elision#RVO), and it's cleaner.

```
SmallVector<Range> cloneOffsetsSizesAndStrides(OpBuilder &builder,
  IREE::Flow::DispatchTensorLoadOp loadOp, SmallVector<Value> &dynamicDims).
```

We can have some helper function that construct `SmallVector<Range>` given vector of `sizes`, vector of `offsets` and vector of `strides`.

If we have concerns that `dynamicDims` is passing by reference, we can at least do

```
void cloneOffsetsSizesAndStrides(OpBuilder &builder,
  IREE::Flow::DispatchTensorLoadOp loadOp,
  SmallVector<Range> &ranges, SmallVector<Value> &dynamicDims).
```

**Review by hanhanW — CHANGES_REQUESTED** [review]
The LLVM programming manual states that we prefer to use ArrayRef<T> or SmallVectorImpl<T> as a parameter type. So maybe use `SmallVectorImpl<T> &` for those parameters.

https://llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h

**Comment by MaheshRavishankar** [comment]
> The LLVM programming manual states that we prefer to use ArrayRef or SmallVectorImpl as a parameter type. So maybe use `SmallVectorImpl<T> &` for those parameters.
> 
> [llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h](https://llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h)

We have discussed this before. I think this predates `SmallVector<T>` (i.e. default N value). So then it could end up with a copy if `N` didnt match.... but since there is a default `N` I dont see a reason to stick to `SmallVectorImpl`. `SmallVector<T>` already means there is no copy, etc. I would call those recommendations outdated.

File: compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp (hunk 2 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp:95]: If you pass `Range`, then you have to first pack everything into a `Range` only to unpack it to get the `SmallVector<OpFoldResult>`  A function with 6 args is not unreasonable. We have `build` methods with 11 arguments, etc.

Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp:95]: > If you pass Range, then you have to first pack everything into a Range only to unpack it to get the SmallVector<OpFoldResult>

yes, and that could be done by creating a helper function.

> A function with 6 args is not unreasonable. We have build methods with 11 arguments, etc.

This is one of parts that I don't like in MLIR -- a builder methods with many arguments, and the order matters. I'm trying to reduce too many arguments functions when possible. This is what I learnt from https://abseil.io/tips/173

> For functions which take multiple arguments which may be confused by the caller or where you want to specify default arguments without having to worry about the order, strongly consider using option structs to increase both convenience and code clarity.

I'm open to both options, just wanna raise a signal about it and think if we've considered it.

**Review by hanhanW — APPROVED** [review]

**Review by hanhanW — APPROVED** [review]

**Pull request merged**